File: app/src/main/java/com/mhacks/android/ui/MainActivity.java
package com.mhacks.android.ui;

import android.app.Fragment;
import android.app.FragmentManager;
import android.app.FragmentTransaction;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.View;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GooglePlayServicesUtil;
import com.google.android.gms.gcm.GoogleCloudMessaging;
import com.mhacks.android.data.model.Token;
import com.mhacks.android.data.network.HackathonCallback;
import com.mhacks.android.data.network.NetworkManager;
import com.mhacks.android.ui.account.AccountFragment;
import com.mhacks.android.ui.announcements.AnnouncementsFragment;
import com.mhacks.android.ui.countdown.CountdownFragment;
import com.mhacks.android.ui.events.ScheduleFragment;
import com.mhacks.android.ui.map.MapViewFragment;
import com.mhacks.android.ui.registration.RegistrationFragment;
import com.mhacks.android.ui.settings.SettingsFragment;
import com.mhacks.android.ui.ticket.TicketFragment;

import org.mhacks.android.R;

import java.io.IOException;
import java.util.Set;

/**
 Activity defines primarily the initial network calls to GCM as well as handle Fragment transactions.
 **/

public class MainActivity extends AppCompatActivity implements ActivityCompat.OnRequestPermissionsResultCallback {
    public static final String TAG = "MainActivity";

    // Permissions
    public static final int LOCATION_REQUEST_CODE = 7;
    String regid;
    String PROJECT_NUMBER;
    String notif;
    // Toolbar
    private Toolbar mToolbar;
    private boolean val;
    //Fragments
    private CountdownFragment countdownFragment;
    private AnnouncementsFragment announcementsFragment;
    private ScheduleFragment scheduleFragment;
    private SettingsFragment settingsFragment;
    private MapViewFragment mapViewFragment;
    private RegistrationFragment registrationFragment;
    private TicketFragment ticketFragment;
    private AccountFragment accountFragment;
    //GCM
    private GoogleCloudMessaging gcm;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);

        // Add the toolbar
        mToolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(mToolbar);

        // If Activity opened from push notification, value will reflect fragment that will initially open
        notif = getIntent().getStringExtra("notif_link");
        PROJECT_NUMBER = getString(R.string.gcm_server_id);

        //Instantiate fragments
        countdownFragment = new CountdownFragment();
//        announcementsFragment = new AnnouncementsFragment();
//        scheduleFragment = new ScheduleFragment();
//        settingsFragment = new SettingsFragment();
//        mapViewFragment = new MapViewFragment();
//        registrationFragment = new RegistrationFragment();
//        ticketFragment = new TicketFragment();
//        accountFragment = new AccountFragment();
//
//        updateFragment(countdownFragment);
        updateGcm();

        if (notif != null) {
            // Opens Announcements
            if (notif.equals("Announcements")){
                updateFragment(announcementsFragment);
            }
        }
    }

//    public void login() {
//        // Log in if possible.
//        SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(this);
//        final String username = sharedPref.getString(SettingsFragment.USERNAME_KEY, "");
//        String password = sharedPref.getString(SettingsFragment.PASSWORD_KEY, "");
//
//        final NetworkManager networkManager = NetworkManager.getInstance();
//        if (username.length() != 0 && password.length() != 0) {
//            networkManager.login(username, password, new HackathonCallback<User>() {
//                @Override
//                public void success(User response) {
//                    userProfile.withName(response.getName());
//                    mDrawer.updateItem(userProfile);
//                }
//
//                @Override
//                public void failure(Throwable error) {
//                }
//            });
//        }
//    }

    public void updateGcm(){
        if (!checkPlayServices()) {
            Log.e(TAG, "No valid Google Play Services APK found.");
            return;
        }
        // Grabs the Google Cloud Messaging REG ID and sends it to the backend

        new AsyncTask<Void, Void, String>() {
            @Override
            protected String doInBackground(Void... params) {
                String msg = "";
                try {
                    if (gcm == null) {
                        gcm = GoogleCloudMessaging.getInstance(getApplicationContext());
                    }
                    // reg_id
                    Bundle data = new Bundle();

                    regid = gcm.register(PROJECT_NUMBER);

                    data.putString("regid",regid);
                    /*InstanceID instanceID = InstanceID.getInstance(getApplicationContext());
                    String token = instanceID.getToken(getString(R.string.gcm_defaultSenderId),
                            GoogleCloudMessaging.INSTANCE_ID_SCOPE, null);*/

                    final SharedPreferences sharedPref = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
                    String gcmPush = sharedPref.getString("gcm", "");
                    Set<String> channels =  sharedPref.getStringSet(SettingsFragment.PUSH_NOTIFICATION_CHANNELS, null);

                    int pref = 1;
                    if (channels != null) {
                        String[] channelPrefs = channels.toArray(new String[channels.size()]);
                        for (int i = 0; i < channelPrefs.length; ++i) {
                            pref += Integer.parseInt(channelPrefs[i]);
                        }
                    } else pref = 63;

                    final Token token = new Token(regid);
                    token.setName(String.valueOf(pref));
                    // active=true by default

                    final NetworkManager networkManager = NetworkManager.getInstance();
                    networkManager.sendToken(token, new HackathonCallback<Token>() {
                        @Override
                        public void success(Token response) {
                            Log.d(TAG, "gcm sent successfully: " + token.getRegistrationId());
                            sharedPref.edit().putString("gcm", regid).apply();
                        }

                        @Override
                        public void failure(Throwable error) {
                            Log.e(TAG, "gcm didnt work", error);
                        }
                    });

                    msg = "Device registered, reg id =" + regid;
                    Log.i("GCM",  msg);

                } catch (IOException ex) {
                    msg = "Error :" + ex.getMessage();
                    Log.e(TAG, "IOException when registering the device", ex);

                }
                return msg;
            }

            @Override
            protected void onPostExecute(String msg) {

            }
        }.execute(null, null, null);
    }


//        // Configure item selection listener
//        mDrawer.setOnDrawerItemClickListener(new Drawer.OnDrawerItemClickListener() {
//            @Override
//            public boolean onItemClick(View view, int position, IDrawerItem drawerItem) {
//                Log.d(TAG, "nav position: " + position);
//                mDrawer.closeDrawer();
//
//                NetworkManager manager = NetworkManager.getInstance();
//                User user =  manager.getCurrentUser();
//
//                // switch 'i' aka position of item
//                // indexing starts at 1 for some reason... probably because of the account header
//                switch (position) {
//                    case 1:
//                        updateFragment(countdownFragment, true);
//                        break;
//                    case 2:
//                        updateFragment(announcementsFragment, true);
//                        break;
//                    case 3:
//                        updateFragment(scheduleFragment, true);
//                        break;
//                    case 4:
//                        requestAndEnableLocation();
//                        break;
//                    case 5:
//                        if (user != null) updateFragment(registrationFragment, true);
//                        else requestLogin();
//                        break;
//                    case 6:
//                        if (user != null) updateFragment(ticketFragment, true);
//                        else requestLogin();
//                        break;
//                    case 8:
//                        updateFragment(accountFragment, true);
//                        break;
//                    case 9:
//                        updateFragment(settingsFragment, true);
//                        break;
//                    default:
//                        return false;
//                }
//
//                return true;
//            }
//        });
//
//    }
//


    // Checks if the user can obtain the correct Google Play Services number
    private boolean checkPlayServices() {
        val = true;
        new AsyncTask<Void, Void, String>() {
            @Override
            protected String doInBackground(Void... params) {
                String msg = "";
                int resultCode = GooglePlayServicesUtil.isGooglePlayServicesAvailable(getApplicationContext());
                if (resultCode != ConnectionResult.SUCCESS)

                {
                    if (GooglePlayServicesUtil.isUserRecoverableError(resultCode)) {
                        //GooglePlayServicesUtil.getErrorDialog(resultCode, this ,GooglePlayServicesUtil.GOOGLE_PLAY_SERVICES_VERSION_CODE).show();
                    } else {
                        Log.e(TAG, "This device is not supported.");
                        finish();
                    }

                    return "false";
                }

                return "true";
            }
            @Override
            protected void onPostExecute(String msg) {
                val = (msg=="true");
            }
        }.execute(null, null, null);
        return val;
    }

    /**
     * Updates the main_fragment_container with the given fragment.
     * @param fragment fragment to replace the main container with
     */
    private void updateFragment(Fragment fragment) {
        if (fragment == null) return; // only used for pre-release while fragments are not finalized

        FragmentManager fragmentManager = getFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
        fragmentTransaction.replace(R.id.main_container, fragment);
        fragmentTransaction.commit();
    }

//    private void requestAndEnableLocation() {
//        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) !=
//            PackageManager.PERMISSION_GRANTED) {
//
//            new AlertDialog.Builder(this)
//                    .setTitle("Location Permission")
//                    .setMessage("The MHacks app uses your location to show you where you are on the map and help you find rooms. Would you like to enable location services?")
//                    .setPositiveButton("Yes", new DialogInterface.OnClickListener() {
//                        public void onClick(DialogInterface dialog, int which) {
//                            // Request permission
//                            ActivityCompat.requestPermissions(MainActivity.this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION,
//                                                                                          Manifest.permission.ACCESS_COARSE_LOCATION},
//                                                              LOCATION_REQUEST_CODE);
//                        }
//                    })
//                    .setNegativeButton("No", new DialogInterface.OnClickListener() {
//                        public void onClick(DialogInterface dialog, int which) {
//                            updateFragment(mapViewFragment, true);
//                        }
//                    })
//                    .show();
//        } else {
//            updateFragment(mapViewFragment, true);
//        }
//    }


    /**
     * Handles all the clicks for the ScheduleFragment and it's fragments.
     * @param v clicked View
     */
    public void scheduleFragmentClick(View v) {
        if (v.getId() == R.id.event_close_button) scheduleFragment.scheduleFragmentClick(v);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode,
                                           @NonNull String[] permissions,
                                           @NonNull int[] grantResults) {
        Log.d(TAG, "ayy");
        switch (requestCode) {
            case LOCATION_REQUEST_CODE:
                if (permissions.length > 0 && (grantResults[0] == PackageManager.PERMISSION_GRANTED || grantResults[1] == PackageManager.PERMISSION_GRANTED)) {
                    Log.d(TAG, "Location permissions granted");
                }
                updateFragment(mapViewFragment);
                break;
        }
    }
}

File: app/src/main/java/com/mhacks/android/ui/MainActivity.kt
package com.mhacks.android.ui

import android.app.FragmentTransaction
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.PorterDuff
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.support.v4.content.ContextCompat
import android.support.v7.widget.Toolbar
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Spinner
import com.mhacks.android.MHacksApplication
import com.mhacks.android.dagger.component.HackathonComponent
import com.mhacks.android.data.kotlin.*
import com.mhacks.android.data.network.fcm.RegistrationIntentService
import com.mhacks.android.data.network.services.HackathonApiService
import com.mhacks.android.data.room.MHacksDatabase
import com.mhacks.android.ui.announcement.AnnouncementFragment
import com.mhacks.android.ui.common.BaseActivity
import com.mhacks.android.ui.common.BaseFragment
import com.mhacks.android.ui.common.NavigationColor
import com.mhacks.android.ui.countdown.WelcomeFragment
import com.mhacks.android.ui.events.EventsFragment
import com.mhacks.android.ui.login.LoginActivity
import com.mhacks.android.ui.map.MapViewFragment
import com.mhacks.android.ui.ticket.TicketDialogFragment
import com.mhacks.android.ui.common.util.GooglePlayUtil
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*
import org.mhacks.android.R
import java.util.concurrent.TimeUnit
import javax.inject.Inject


/**
 * Activity defines primarily the initial hackathonService calls to GCM as well as handle Fragment transactions.
 */
class MainActivity : BaseActivity(),
        ActivityCompat.OnRequestPermissionsResultCallback,
        BaseFragment.OnNavigationChangeListener,
        WelcomeFragment.Callback,
        TicketDialogFragment.Callback,
        MapViewFragment.Callback,
        AnnouncementFragment.Callback,
        EventsFragment.Callback {

    // Callbacks to properties and methods on the application class.
    private val appCallback by lazy {
        application as MHacksApplication
    }

    private var notif: String? = null

//    lateinit var regid: String
//    val PROJECT_NUMBER: String by lazy { getString(R.string.gcm_server_id) }

    @Inject lateinit var hackathonService: HackathonApiService
    @Inject lateinit var mhacksDatabase: MHacksDatabase
    private lateinit var menuItem: MenuItem

    private lateinit var navigationSpinner: Spinner

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        if (GooglePlayUtil.checkPlayServices(this)) {
            val intent = Intent(this, RegistrationIntentService::class.java)
            startService(intent)
        }

        appCallback.hackathonComponent.inject(this)
        setTheme(R.style.MHacksTheme)
        checkIfLogin()
    }


    override fun fetchFloors(success: (floor: List<Floor>) -> Unit,
                             failure: (error: Throwable) -> Unit) {
        checkIfNetworkIsPresent(this, {
            hackathonService.getMetaFloors()
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            { response -> success(response.floors) },
                            { error -> failure(error) })
        })

    }



    override fun checkOrFetchConfig(success: (config: MetaConfiguration) -> Unit,
                                   failure: (error: Throwable) -> Unit) {
        checkIfNetworkIsPresent(this,
            { hackathonService.getMetaConfiguration()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { response -> success(response) },
                        { error -> failure(error) })
            })
    }


    private fun checkIfLogin() {
        mhacksDatabase.loginDao().getLogin()
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { login ->
                            appCallback.setAuthInterceptorToken(login.token)
                            initActivity() },
                        { startLoginActivity() }
                )
    }

    override fun fetchAnnouncements(success: (announcements: List<Announcements>) -> Unit,
                                    failure: (error: Throwable) -> Unit) {
        checkIfNetworkIsPresent(this, {
            Observable.timer(3000, TimeUnit.MILLISECONDS)
                    .startWith(0)
                    .flatMap {
                        hackathonService.getMetaAnnouncements()
                    }
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            { (announcements) -> success(announcements) },
                            { error -> failure(error) })
        })
    }

    private fun showTicketDialogFragment() {
        val ft = supportFragmentManager.beginTransaction()
        val prev = supportFragmentManager.findFragmentByTag("dialog")
        if (prev != null) {
            ft.remove(prev)
        }
        ft.addToBackStack(null)

        val ticket: TicketDialogFragment = TicketDialogFragment
                .newInstance()
        ticket.show(ft, "dialog")
    }

    private fun updateFragment(fragment: Fragment?) {
        if (fragment == null) return  // only used for pre-release while fragments are not finalized

        supportFragmentManager.beginTransaction()
                .setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
                .replace(R.id.fragment_container, fragment)
                .commit()
    }

    override fun startLoginActivity() {
        startActivity(Intent(this, LoginActivity::class.java))
        finish()
    }

//    private fun startQRScanActivity() {
//        startActivity(Intent(this, QRScanActivity::class.java))
//        finish()
//    }

    override fun checkOrFetchUser(success: (user: User) -> Unit,
                                  failure: (error: Throwable) -> Unit) {
        mhacksDatabase.userDao().getUser()
                .onErrorResumeNext ({
                    mhacksDatabase.loginDao().getLogin()
                            .flatMap({ login ->
                                appCallback.setAuthInterceptorToken(login.token)
                                hackathonService.getMetaUser()
                                        .flatMap { user -> Single.just(user.user) }
                            })
                })
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { response -> success(response) },
                        { error -> failure(error) })
    }

    override fun fetchEvents(success: (events: List<Events>) -> Unit,
                             failure: (error: Throwable) -> Unit) {
        checkIfNetworkIsPresent(this, {
            hackathonService.getMetaEvent()
                    .subscribeOn(Schedulers.newThread())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            { (events) -> success(events) },
                            { error -> failure(error) })
        })
    }

    override fun updateFloors(floors: List<Floor>, listener: OnItemSelectedListener) {

        val adapter = ArrayAdapter<Floor>(this, R.layout.floors_spinner_item, floors)
        adapter.setDropDownViewResource(R.layout.map_view_dropdown_item_1line)
        navigationSpinner.adapter = adapter

        navigationSpinner.onItemSelectedListener = listener
    }


    override fun showFloorOptions() {
       // navigationSpinner.visibility = View.VISIBLE
    }

    override fun hideFloorOptions() {
        navigationSpinner.visibility = View.GONE
    }


    override fun onRequestPermissionsResult(requestCode: Int,
                                            permissions: Array<String>,
                                            grantResults: IntArray) {
        when (requestCode) {
            LOCATION_REQUEST_CODE ->
                if (permissions.isNotEmpty() &&
                        (grantResults[0] == PackageManager.PERMISSION_GRANTED ||
                         grantResults[1] == PackageManager.PERMISSION_GRANTED)) {
            }
        }
    }

    private fun initActivity() {

        setSystemFullScreenUI()
        setContentView(R.layout.activity_main)
        setBottomNavigationColor(
                NavigationColor(R.color.colorPrimary, R.color.colorPrimaryDark))

        qr_ticket_fab.setOnClickListener({
            showTicketDialogFragment()
        })

        menuItem = navigation.menu.getItem(0)
        menuItem.setTitle(R.string.title_home)

        setSupportActionBar(toolbar)

        addMapFloorSpinner()

        toolbar.addView(navigationSpinner, Toolbar.LayoutParams(Gravity.END))

        updateFragment(WelcomeFragment.instance)



        // If Activity opened from push notification, value will reflect fragment that will initially open
        notif = intent.getStringExtra("notif_link")



        navigation?.setOnNavigationItemSelectedListener({ item ->
            when (item.itemId) {
                R.id.navigation_home -> updateFragment(WelcomeFragment.instance)
                R.id.navigation_announcements -> updateFragment(AnnouncementFragment.instance)
                R.id.navigation_events -> updateFragment(EventsFragment.instance)
                R.id.navigation_map -> updateFragment(MapViewFragment.instance)
            }
            menuItem = item
            true
        })
    }

    private fun addMapFloorSpinner() {

        navigationSpinner = Spinner(supportActionBar?.themedContext)

        navigationSpinner.visibility = View.GONE

        navigationSpinner.background.setColorFilter(
                ContextCompat.getColor(this, R.color.white),
                PorterDuff.Mode.SRC_ATOP)

    }

    interface OnFromMainActivityCallback {

        fun setAuthInterceptorToken(token: String)

        val hackathonComponent: HackathonComponent
    }

    companion object {

        val LOCATION_REQUEST_CODE = 7
    }
}
